### PR TITLE
[BEAM-7194] EXCEPT DISTINCT behavior when right set contains a value is incorrect

### PR DESCRIPTION
**Pull request by TheNeuralBit**
EXCEPT DISTINCT now yields a particular value if and only if the left multiset contains `m > 0` instances of that value, and the right multiset contains none.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by TheNeuralBit** [comment]
R: @amaliujia, @apilloud

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java [hunk 1]
```diff
@@ -97,8 +97,10 @@ public void processElement(ProcessContext ctx) {
           }
           break;
         case MINUS:
-          // Say for Row R, there are m instances on left and n instances on right,
-          // EXCEPT ALL outputs MAX(m - n, 0) instances of R.
+          // Say for Row R, there are m instances on left and n instances on right:
+          // - EXCEPT ALL outputs MAX(m - n, 0) instances of R.
+          // - EXCEPT DISTINCT and EXCEPT output a single instance of R if m > 0 and n == 0, else
```
Comment (amaliujia) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: According to Calcite SQL doc( https://calcite.apache.org/docs/reference.html), 

ALL is the default for UNION/INTERSECT/DISTINCT. So you can write something like
```
EXCEPT [ALL] 
EXCEPT DISTINCT
```

Comment (TheNeuralBit) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: Ah good point. That seems to be at odds with the test case I modified though - the SQL query in `testExcept` just says `EXCEPT` and seems to run with `all=false`. Any idea what's going on there?

Comment (amaliujia) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: That's interesting. I am thinking if Calcite doc is stale. 

Could you check the logical plan and see what's default setting is?

Comment (TheNeuralBit) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamSetOperatorsTransforms.java]: Discussed this a bit offline. It looks like the default in Calcite [is in fact `EXCEPT DISTINCT`](https://github.com/apache/calcite/blob/master/core/src/main/codegen/templates/Parser.jj#L6011), and it's actually not mentioned on the calcite reference page. The line we were both referencing says that `MULTISET EXCEPT` defaults to all. It doesn't seem to specify `EXCEPT`'s default behavior anywhere.

I'm going to change this comment to `EXCEPT ALL` and `EXCEPT [DISTINCT]` and resolve this.

**Comment by TheNeuralBit** [comment]
I'm not sure what's going wrong with the pre commits here. The test failure seems to be in `org.apache.beam.sdk.scheas.transforms.CoGroupTest` and doesn't seem related to this change.

In fact even if I run `./gradlew :beam-runners-direct-java:needsRunnerTests` on origin/master it sill fails.

**Comment by amaliujia** [comment]
Did you rebase to head of upstream/master?

**Comment by TheNeuralBit** [comment]
It looks like actually the `CoGroupTest` failures were unrelated to the CI failure, they were due to a quirk of the environment where I was running the test (which, happily, helped me discover the issue that #8485 fixes).

I've rebased and forced push, hopefully this gets a green run now.

**Comment by TheNeuralBit** [comment]
@apilloud would you be ok merging this?

**Comment by apilloud** [comment]
LGTM.

**Pull request merged**